File: vripper-server/src/main/java/tn/mnlr/vripper/host/ImageZillaHost.java
package tn.mnlr.vripper.host;

import lombok.extern.slf4j.Slf4j;
import org.apache.http.client.protocol.HttpClientContext;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import tn.mnlr.vripper.exception.HostException;
import tn.mnlr.vripper.exception.XpathException;
import tn.mnlr.vripper.services.HostService;
import tn.mnlr.vripper.services.XpathService;

@Service
@Slf4j
public class ImageZillaHost extends Host {

  private static final String host = "imagezilla.net";
  private static final String lookup = "imagezilla.net/show";
  private static final String IMG_XPATH = "//img[@id='photo']";

  private final HostService hostService;
  private final XpathService xpathService;

  @Autowired
  public ImageZillaHost(HostService hostService, XpathService xpathService) {
    this.hostService = hostService;
    this.xpathService = xpathService;
  }

  @Override
  public String getHost() {
    return host;
  }

  @Override
  public String getLookup() {
    return lookup;
  }

  @Override
  public HostService.NameUrl getNameAndUrl(final String url, final HttpClientContext context)
      throws HostException {

    Document doc = hostService.getResponse(url, context).getDocument();

    String title;
    try {
      log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, url));
      Node titleNode = xpathService.getAsNode(doc, IMG_XPATH).getAttributes().getNamedItem("title");
      log.debug(String.format("Resolving name for %s", url));
      if (titleNode != null) {
        title = titleNode.getTextContent().trim();
      } else {
        title = null;
      }
    } catch (XpathException e) {
      throw new HostException(e);
    }

    if (title == null || title.isEmpty()) {
      title = hostService.getDefaultImageName(url);
    }

    try {
      return new HostService.NameUrl(title, url.replace("show", "images"));
    } catch (Exception e) {
      throw new HostException("Unexpected error occurred", e);
    }
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/host/ImageZillaHost.kt
package me.mnlr.vripper.host

import org.w3c.dom.Document
import me.mnlr.vripper.delegate.LoggerDelegate
import me.mnlr.vripper.download.ImageDownloadContext
import me.mnlr.vripper.exception.HostException
import me.mnlr.vripper.exception.XpathException
import me.mnlr.vripper.services.*

class ImageZillaHost(
    httpService: HTTPService,
    dataTransaction: DataTransaction,
    globalStateService: GlobalStateService,
) : Host("imagezilla.net", httpService, dataTransaction, globalStateService) {
    private val log by LoggerDelegate()

    @Throws(HostException::class)
    override fun resolve(
        url: String,
        document: Document,
        context: ImageDownloadContext
    ): Pair<String, String> {
        val titleNode = try {
            log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, url))
            XpathService.getAsNode(document, IMG_XPATH)
        } catch (e: XpathException) {
            throw HostException(e)
        } ?: throw HostException(
            String.format(
                "Xpath '%s' cannot be found in '%s'",
                IMG_XPATH,
                url
            )
        )
        log.debug(String.format("Resolving name for %s", url))
        var title = titleNode.attributes.getNamedItem("title").textContent.trim()
        titleNode.textContent.trim()
        if (title.isEmpty()) {
            title = getDefaultImageName(url)
        }
        return try {
            Pair(title, url.replace("show", "images"))
        } catch (e: Exception) {
            throw HostException("Unexpected error occurred", e)
        }
    }

    companion object {
        private const val IMG_XPATH = "//img[@id='photo']"
    }
}